service/history/workflow/workflow_task_state_machine.go

```diff
@@ -447,15 +451,32 @@ func (m *workflowTaskStateMachine) AddWorkflowTaskStartedEvent(
 
 	m.emitWorkflowTaskAttemptStats(workflowTask.Attempt)
 
-	// TODO merge active & passive task generation
-	if err := m.ms.taskGenerator.GenerateStartWorkflowTaskTasks(
-		scheduledEventID,
-	); err != nil {
-		return nil, nil, err
+	// Always bypass task generation for speculative WT.
+	if workflowTask.Type != enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE {
+		// TODO merge active & passive task generation
+		if err := m.ms.taskGenerator.GenerateStartWorkflowTaskTasks(
+			scheduledEventID,
+		); err != nil {
+			return nil, nil, err
+		}
 	}
+
 	return startedEvent, workflowTask, err
 }
+func (m *workflowTaskStateMachine) skipWorkflowTaskCompleted(workflowTaskType enumsspb.WorkflowTaskType, request *workflowservice.RespondWorkflowTaskCompletedRequest) bool {
+	if workflowTaskType != enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE || len(request.GetCommands()) != 0 {
```

Comment:
FWIW, this function actually gets easier if we end up going down the ProtocolCommand route.